### ISSUE TITLE
What is the difference between a `toml` and a `ConfigFile`?

### ISSUE THREAD

**Issue by ThePat02** [Dev-Marco/Godot-TOML#1]
Godot already supports the parsing of [ConfigFiles](https://docs.godotengine.org/en/stable/classes/class_configfile.html), which at a quick glance look basically the same. Am I missing something here?

**Comment by Dev-Marco** [comment]
Glad you asked! 😊

It’s true that **TOML** and **ConfigFiles**, implemented as a pseudo-INI format, look quite similar. In the context of the Godot Engine and within your game, you might prefer using ConfigFiles because they allow you to save and load engine-specific data types. However, one issue with these pseudo-INI formats, as well as the INI format itself, is that they lack a formal standard. Not all INI parsers handle the same set of features consistently. TOML, on the other hand, has a well-defined standard, which ensures consistent parsing across different tools.

In ConfigFiles, every key-value pair must belong to a section, unlike in TOML, where key-value pairs can be placed in the root of the file. To give an analogy from the world of JSON, this would mean that every section in ConfigFiles is like an object in a list with its key-value pairs, while the root is simply a list. In TOML, the root is an object itself, allowing you to place both key-value pairs and sections (which are also objects) directly within it.

Here’s an example:

❌ invalid **ConfigFiles**
✅ valid **TOML**
```toml
key_without_secion = value

[section]
key = value
```

Another key feature of **TOML** is its support for nested structures:
```toml
[section]
key = value

[section.subsection]
key2= value2
```

In **JSON**, this would translate to:
```json
{
    "section": {
        "key": "value",
        "subsection": {
            "key2": "value2"
        }
    }
}
```
While some INI parsers support one level of subsections, I’m fairly certain that Godot’s ConfigFile doesn’t. TOML, however, allows unlimited nesting, such as `[section.subsection_1.subsection_2.subsection_x]`. Additionally, if you wrap a section in double brackets `[[section.sublist]]`, it creates a list in TOML (and this works with nesting too).

Another thing with ConfigFiles is that, in my opinion, they are more intended for saving and loading data related to your game that is used within the session of the game. And because there might be engine specific data types inside your ConfigFile, you shouldn’t manually edit the text inside them unless you really know what you’re doing, as you could easily break the state of saved game objects. TOML, on the other hand, doesn’t deal with engine-specific types but rather provides a compact set of general data types. This allows you to structure data in a way that’s flexible and human-readable, without the risk of unintentionally disrupting complex engine states.

In summary, TOML is a standardized format for writing configuration files, but it can also be used like JSON as a data exchange format that’s understood by many applications across various systems. It’s simple yet expressive, and unlike YAML, it’s not as difficult to parse.
